[analyzer] Crash with "ArrayInitLoopExpr contains unexpected source expression"

Got an error while running clang-tidy. Clang-tidy is executed by CodeChecker, flags and options are generated by the cmake with -DCMAKE_EXPORT_COMPILE_COMMANDS=1. The same error occurs when CodeChecker runs clang static analyzer

```
Stack dump:
0.	Program arguments: /usr/lib/llvm-18/bin/clang-tidy /home/waste/Projects/cplusplus/algorithms/tf_idf_actrie/test_actrie.cpp --export-fixes /home/waste/Projects/cplusplus/algorithms/tests/cmake-build-codechecker/report_clang/fixit/test_actrie.cpp_clang-tidy_889ac94762fd440ca0328d94131f3632.yaml -- -Qunused-arguments -Wno-everything -x c++ --target=x86_64-pc-linux-gnu -D_GLIBCXX_CONCEPT_CHECKS -D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC -D_GLIBCXX_SANITIZE_VECTOR -D_LIBCPP_ENABLE_ASSERTIONS -fsanitize=address,undefined -O2 -g -std=c++20 -Wall -Wextra -Wcast-qual -Wpedantic -Wunused -Wshadow -Wnull-dereference -Wundef -Wwrite-strings -Wsign-conversion -Wsign-compare -Wsign-promo -Wconversion -Wmissing-noreturn -Wunreachable-code -Wcast-align -Wformat=2 -Wswitch-bool -Wswitch-default -Wswitch-enum -Wold-style-cast -Wdeprecated -Weffc++ -fstack-protector-strong -mshstk -Warray-bounds -Wshift-overflow -Wshift-sign-overflow -Wshorten-64-to-32 -Wthread-safety -fcolor-diagnostics -fansi-escape-codes -ftemplate-backtrace-limit=0 -U_GLIBCXX_USE_DEPRECATED -isystem /usr/include/c++/13 -isystem /usr/include/x86_64-linux-gnu/c++/13 -isystem /usr/include/c++/13/backward -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include
1.	<eof> parser at end of file
2.	While analyzing stack: 
	#0 Calling (anonymous namespace)::replacing_actrie_tests::test_replacing_actrie(const std::string_view (&)[5][2], std::string &, const std::string_view, const _Bool) at line /home/waste/Projects/cplusplus/algorithms/tf_idf_actrie/test_actrie.cpp:155:5 <Spelling=/home/waste/Projects/cplusplus/algorithms/tf_idf_actrie/test_actrie.cpp:155:12>
	#1 Calling (anonymous namespace)::replacing_actrie_tests::test0() at line 448
	#2 Calling main()
3.	/home/waste/Projects/cplusplus/algorithms/tf_idf_actrie/test_actrie.cpp:118:44: Error evaluating statement
4.	/home/waste/Projects/cplusplus/algorithms/tf_idf_actrie/test_actrie.cpp:118:44: Error evaluating statement
 #0 sys::PrintStackTrace(raw_ostream&, int) Signals.inc:723:13
 #1 sys::RunSignalHandlers() Signals.cpp:106:18
 #2 SignalHandler Signals.inc:413:1
 #3 (/lib/x86_64-linux-gnu/libc.so.6+0x45320)
 #4 getCanonicalDecl clang/AST/Decl.h:1274:40
 #5 MemRegionManager::getVarRegion(VarDecl const*, LocationContext const*) MemRegion.cpp:1007:10
 #6 clang/StaticAnalyzer/Core/PathSensitive/Store.h:136:5
 #7 getType clang/AST/Expr.h:143:37
 #8 bindRequiredArrayElementToEnvironment ExprEngineCXX.cpp:577:39
 #9 ExprEngine::handleConstructor(Expr const*, ExplodedNode*, ExplodedNodeSet&) ExprEngineCXX.cpp:660:15
#10 ExprEngine::Visit(Stmt const*, ExplodedNode*, ExplodedNodeSet&) ExprEngine.cpp:0:7
#11 ExprEngine::ProcessStmt(Stmt const*, ExplodedNode*) ExprEngine.cpp:1132:9
#12 ExprEngine::processCFGElement(CFGElement, ExplodedNode*, unsigned int, NodeBuilderContext*) ExprEngine.cpp:0:7
#13 CoreEngine::dispatchWorkItem(ExplodedNode*, ProgramPoint, WorkListUnit const&) CoreEngine.cpp:223:1
```

Result of the `clang --analyze file.cpp` is pretty the same when using clang 20.0.0 built from the main branch:

PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /mnt/wd15/projects/c++/llvm-project/build/bin/clang --analyze file.cpp
1.      <eof> parser at end of file
2.      While analyzing stack: 
        #0 Calling main()
3.      file.cpp:7:32: Error evaluating statement
4.      file.cpp:7:32: Error evaluating statement
 #0 0x00005b5c8701c6a2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x720e6a2)
 #1 0x00005b5c8701a3ec llvm::sys::CleanupOnSignal(unsigned long) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x720c3ec)
 #2 0x00005b5c86f53058 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x000071a780045320 (/lib/x86_64-linux-gnu/libc.so.6+0x45320)
 #4 0x00005b5c8969e840 clang::ento::MemRegionManager::getVarRegion(clang::VarDecl const*, clang::LocationContext const*) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x9890840)
 #5 0x00005b5c8966aac9 clang::ento::ExprEngine::handleConstructor(clang::Expr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x985cac9)
 #6 0x00005b5c8963f1f2 clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x98311f2)
 #7 0x00005b5c896411fd clang::ento::ExprEngine::ProcessStmt(clang::Stmt const*, clang::ento::ExplodedNode*) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x98331fd)
 #8 0x00005b5c8964ee0a clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x9840e0a)
 #9 0x00005b5c895fba62 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x97eda62)
#10 0x00005b5c895fbcce clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x97edcce)
#11 0x00005b5c891375fa (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*, void>>*) AnalysisConsumer.cpp:0:0
#12 0x00005b5c8914bfd3 (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) AnalysisConsumer.cpp:0:0
#13 0x00005b5c8975b42c clang::ParseAST(clang::Sema&, bool, bool) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x994d42c)
#14 0x00005b5c87ed53b9 clang::FrontendAction::Execute() (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x80c73b9)
#15 0x00005b5c87e55d9b clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x8047d9b)
#16 0x00005b5c87fb949b clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x81ab49b)
#17 0x00005b5c843c63c1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x45b83c1)
#18 0x00005b5c843bda4a ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#19 0x00005b5c87c3ca0d void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#20 0x00005b5c86f534a7 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x71454a7)
#21 0x00005b5c87c3d0d1 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#22 0x00005b5c87c009ba clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x7df29ba)
#23 0x00005b5c87c01a6d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x7df3a6d)
#24 0x00005b5c87c1422c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x7e0622c)
#25 0x00005b5c843c2116 clang_main(int, char**, llvm::ToolContext const&) (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x45b4116)
#26 0x00005b5c843d479a main (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x45c679a)
#27 0x000071a78002a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#28 0x000071a78002a28b call_init ./csu/../csu/libc-start.c:128:20
#29 0x000071a78002a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#30 0x00005b5c843bd4a5 _start (/mnt/wd15/projects/c++/llvm-project/build/bin/clang+0x45af4a5)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 20.0.0git (https://github.com/llvm/llvm-project.git c77d8edf80570f450122a7687100553836149652)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/wd15/projects/c++/llvm-project/build/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/file-26c5dd.cpp
clang: note: diagnostic msg: /tmp/file-26c5dd.sh
clang: note: diagnostic msg: 

********************

--- a/clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp
+++ b/clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp
@@ -513,70 +513,25 @@ ProgramStateRef ExprEngine::updateObjectsUnderConstruction(
 static ProgramStateRef
 bindRequiredArrayElementToEnvironment(ProgramStateRef State,
                                       const ArrayInitLoopExpr *AILE,
-                                      const LocationContext *LCtx, SVal Idx) {
-  // The ctor in this case is guaranteed to be a copy ctor, otherwise we hit a
-  // compile time error.
-  //
-  //  -ArrayInitLoopExpr                <-- we're here
-  //   |-OpaqueValueExpr
-  //   | `-DeclRefExpr                  <-- match this
-  //   `-CXXConstructExpr
-  //     `-ImplicitCastExpr
-  //       `-ArraySubscriptExpr
-  //         |-ImplicitCastExpr
-  //         | `-OpaqueValueExpr
-  //         |   `-DeclRefExpr
-  //         `-ArrayInitIndexExpr
-  //
-  // The resulting expression might look like the one below in an implicit
-  // copy/move ctor.
-  //
-  //   ArrayInitLoopExpr                <-- we're here
-  //   |-OpaqueValueExpr
-  //   | `-MemberExpr                   <-- match this
-  //   |  (`-CXXStaticCastExpr)         <-- move ctor only
-  //   |     `-DeclRefExpr
-  //   `-CXXConstructExpr
-  //     `-ArraySubscriptExpr
-  //       |-ImplicitCastExpr
-  //       | `-OpaqueValueExpr
-  //       |   `-MemberExpr
-  //       |     `-DeclRefExpr
-  //       `-ArrayInitIndexExpr
-  //
-  // The resulting expression for a multidimensional array.
-  // ArrayInitLoopExpr                  <-- we're here
-  // |-OpaqueValueExpr
-  // | `-DeclRefExpr                    <-- match this
-  // `-ArrayInitLoopExpr
-  //   |-OpaqueValueExpr
-  //   | `-ArraySubscriptExpr
-  //   |   |-ImplicitCastExpr
-  //   |   | `-OpaqueValueExpr
-  //   |   |   `-DeclRefExpr
-  //   |   `-ArrayInitIndexExpr
-  //   `-CXXConstructExpr             <-- extract this
-  //     ` ...
-
-  const auto *OVESrc = AILE->getCommonExpr()->getSourceExpr();
+                                      const LocationContext *LCtx, NonLoc Idx) {
+  SValBuilder &SVB = State->getStateManager().getSValBuilder();
+  MemRegionManager &MRMgr = SVB.getRegionManager();
+  ASTContext &Ctx = SVB.getContext();
 
   // HACK: There is no way we can put the index of the array element into the
   // CFG unless we unroll the loop, so we manually select and bind the required
   // parameter to the environment.
-  const auto *CE =
+  const Expr *SourceArray = AILE->getCommonExpr()->getSourceExpr();
+  const auto *Ctor =
       cast<CXXConstructExpr>(extractElementInitializerFromNestedAILE(AILE));
 
-  SVal Base = UnknownVal();
-  if (const auto *ME = dyn_cast<MemberExpr>(OVESrc))
-    Base = State->getSVal(ME, LCtx);
-  else if (const auto *DRE = dyn_cast<DeclRefExpr>(OVESrc))
-    Base = State->getLValue(cast<VarDecl>(DRE->getDecl()), LCtx);
-  else
-    llvm_unreachable("ArrayInitLoopExpr contains unexpected source expression");
+  const auto *SourceArrayRegion =
+      cast<SubRegion>(State->getSVal(SourceArray, LCtx).getAsRegion());
+  const ElementRegion *ElementRegion =
+      MRMgr.getElementRegion(Ctor->getType(), Idx, SourceArrayRegion, Ctx);
 
-  SVal NthElem = State->getLValue(CE->getType(), Idx, Base);
-
-  return State->BindExpr(CE->getArg(0), LCtx, NthElem);
+  return State->BindExpr(Ctor->getArg(0), LCtx,
+                         loc::MemRegionVal(ElementRegion));
 }
 
 void ExprEngine::handleConstructor(const Expr *E,

--- a/clang/test/Analysis/array-init-loop.cpp
+++ b/clang/test/Analysis/array-init-loop.cpp
@@ -330,3 +330,41 @@ void no_crash() {
 }
 
 } // namespace crash
+
+namespace array_subscript_initializer {
+struct S {
+  int x;
+};
+
+void no_crash() {
+  S arr[][2] = {{1, 2}};
+
+  const auto [a, b] = arr[0]; // no-crash
+
+  clang_analyzer_eval(a.x == 1); // expected-warning{{TRUE}}
+  clang_analyzer_eval(b.x == 2); // expected-warning{{TRUE}}
+}
+} // namespace array_subscript_initializer
+
+namespace iterator_initializer {
+struct S {
+  int x;
+};
+
+void no_crash() {
+  S arr[][2] = {{1, 2}, {3, 4}};
+
+  int i = 0;
+  for (const auto [a, b] : arr) { // no-crash
+    if (i == 0) {
+      clang_analyzer_eval(a.x == 1); // expected-warning{{TRUE}}
+      clang_analyzer_eval(b.x == 2); // expected-warning{{TRUE}}
+    } else {
+      clang_analyzer_eval(a.x == 3); // expected-warning{{TRUE}}
+      clang_analyzer_eval(b.x == 4); // expected-warning{{TRUE}}
+    }
+
+    ++i;
+  }
+}
+} // namespace iterator_initializer

